### PR TITLE
[Frontend] Create login/signup pages

### PR DESCRIPTION
**Pull request by brixtuyor**
# Related Links

- Task Link: https://app.asana.com/0/1202403928837322/1202403928837332/f
- Login Route Link: 127.0.0.1:8000/
- Signup Route Link: 127.0.0.1:8000/signup

# Description

- Created login page
- Created signup page
- Additional packages added

# Notes

- Other validation sub-task will be worked on once the back-end has already been established
- Please run `npm install` to append updates in package.json
- Additionally, run `php artisan migrate` and `php artisan db:seed` to populate users table with single data
 
## Screenshots

image
image

### CODE REVIEW

File: database/factories/UserFactory.php [hunk 1]
```diff
@@ -18,9 +18,9 @@ class UserFactory extends Factory
     public function definition()
     {
         return [
-            'name' => $this->faker->name(),
-            'email' => $this->faker->unique()->safeEmail(),
-            'email_verified_at' => now(),
+            'firstname' => "Brix",
+            'lastname' => "Tuyor",
+            'email' => "brixtuyor@email.com",
             'password' => '$2y$10$92IXUNpkjO0rOQ5byMi.Ye4oKoEa3Ro9llC/.og/at2.uheWG/igi', // password
```
Comment (roseaugusto) [inline database/factories/UserFactory.php:24]: @brixtuyor even in seeders, please use bcrypt for uniformity.

File: resources/css/app.css [hunk 2]
```diff
@@ -1,21 +0,0 @@
-@import url("https://fonts.googleapis.com/css2?family=Nunito&display=swap");
```
Comment (roseaugusto) [inline resources/css/app.css:1]: If not used remove the file completely

File: resources/js/components/App.tsx [hunk 3]
```diff
@@ -1,10 +1,16 @@
...
+
 function App() {
     return (
-        <div className="app">
-            <h1>Hi, my name is Brix!</h1>
-            <p>I am currently building an E-Learning System App.</p>
-            <p>Please standby. It is still in progress...</p>
-        </div>
+        <BrowserRouter>
+            <Routes>
```
Comment (roseaugusto) [inline resources/js/components/App.tsx]: Use a different file for the routes and import here instead.

File: resources/js/components/Login.tsx [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    FormLabel,
+    FormErrorMessage,
+    Input,
+    Button,
+    Text,
+    Link,
+} from "@chakra-ui/react";
+import { useState } from "react";
+
+type LoginInputs = {
```
Comment (roseaugusto) [inline resources/js/components/Login.tsx]: make a folder can be called as /utils/types to organize all types/interfaces

File: resources/js/components/Login.tsx [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        setTimeout(() => {
+            setIsLoading(false);
+        }, 2000);
+
+        setFormInputs((previousFormInputs) => ({
+            ...previousFormInputs,
+            password: "",
+        }));
+
+        console.log(formInputs);
```
Comment (roseaugusto) [inline resources/js/components/Login.tsx]: remove console.log

File: resources/js/components/Signup.tsx [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+            setIsLoading(false);
+        }, 2000);
+
+        setFormInputs((previousFormInputs) => ({
+            ...previousFormInputs,
+            password: "",
+            confirmPassword: "",
+        }));
+
+        console.log(formInputs);
```
Comment (roseaugusto) [inline resources/js/components/Signup.tsx]: remove this as well

File: resources/js/components/Signup.tsx [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    });
+
+    const [isLoading, setIsLoading] = useState(false);
+
+    const onSubmit = (e: React.FormEvent<HTMLFormElement>) => {
+        e.preventDefault();
+
+        setIsLoading(true);
+
+        setTimeout(() => {
```
Comment (roseaugusto) [inline resources/js/components/Signup.tsx]: Bad practice to use setTimeOut, it should be check using the http response. Remove this and refactor it later when it's functional.

File: resources/js/components/Signup.tsx [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                                onChange={onChange}
+                            />
+                        </FormControl>
+                        <Button
+                            type="submit"
+                            w="full"
+                            fontWeight="normal"
+                            isLoading={isLoading}
+                            disabled={
+                                formInputs.firstName.trim() === "" ||
```
Comment (roseaugusto) [inline resources/js/components/Signup.tsx]: if the inline condition is more than 2 lines, make separate function to handle this

File: database/factories/UserFactory.php (hunk 1 above)
Comment (brixtuyor) [reply database/factories/UserFactory.php:24]: Copy that. I just tried it initially. Will change that when I get to the backend part.

File: resources/css/app.css (hunk 2 above)
Comment (brixtuyor) [reply resources/css/app.css:1]: `app.css` from both public and resources file removed

File: resources/js/components/App.tsx (hunk 3 above)
Comment (brixtuyor) [reply resources/js/components/App.tsx]: Refactored it to have a single View component in the App

File: resources/js/components/Login.tsx (hunk 4 above)
Comment (brixtuyor) [reply resources/js/components/Login.tsx]: I will be creating @types folder to make my types global during development mode

File: resources/js/components/Login.tsx (hunk 5 above)
Comment (brixtuyor) [reply resources/js/components/Login.tsx]: Removed

File: resources/js/components/Signup.tsx (hunk 6 above)
Comment (brixtuyor) [reply resources/js/components/Signup.tsx]: Removed

File: resources/js/components/Signup.tsx (hunk 7 above)
Comment (brixtuyor) [reply resources/js/components/Signup.tsx]: Will do, thanks

File: resources/js/components/Signup.tsx (hunk 8 above)
Comment (brixtuyor) [reply resources/js/components/Signup.tsx]: Refactored and used a helper function `isDisabled`

File: resources/js/components/View.tsx [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import { BrowserRouter, Route, Routes } from "react-router-dom";
+
+import Login from "./Login";
+import Signup from "./Signup";
+
+function View() {
+    return (
+        <BrowserRouter>
+            <Routes>
```
Comment (roseaugusto) [inline resources/js/components/View.tsx]:
```suggestion
import Routes from './routes';

            <Router>
               <Routes />
           </Router>
```
What I meant is like this. Just the routes itself, not the the whole View.

Comment (brixtuyor) [reply resources/js/components/View.tsx]: Gotcha, modified.

image

File: resources/js/components/Login.tsx [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+    };
+
+    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        setFormInputs((previousFormInputs) => ({
+            ...previousFormInputs,
+            [e.target.name]: e.target.value,
+        }));
+    };
+
+    const onBlur = (e: React.FocusEvent<HTMLInputElement, Element>) => {
```
Comment (roseaugusto) [inline resources/js/components/Login.tsx]: Put all validations on a separate file /utils/helper

File: resources/js/components/Signup.tsx [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+    };
+
+    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        setFormInputs((previousFormInputs) => ({
+            ...previousFormInputs,
+            [e.target.name]: e.target.value,
+        }));
+    };
+
+    const onBlur = (e: React.FocusEvent<HTMLInputElement, Element>) => {
```
Comment (roseaugusto) [inline resources/js/components/Signup.tsx]: this one too

File: resources/js/utils/validation/doesMatch.ts [hunk 12]
```diff
@@ -0,0 +1,5 @@
+function doesMatch(password: string, confirmPassword: string) {
```
Comment (roseaugusto) [inline resources/js/utils/validation/doesMatch.ts]: put in helper file

File: resources/js/utils/validation/isAlphaNumeric.ts [hunk 13]
```diff
@@ -0,0 +1,5 @@
+function isAlphaNumeric(password: string) {
```
Comment (roseaugusto) [inline resources/js/utils/validation/isAlphaNumeric.ts]: put in helper file

File: resources/js/utils/validation/isMinimum.ts [hunk 14]
```diff
@@ -0,0 +1,5 @@
+function isMinimum(value: string, minLength: number = 8) {
```
Comment (roseaugusto) [inline resources/js/utils/validation/isMinimum.ts]: put in helper file

File: resources/js/utils/validation/isRequired.ts [hunk 15]
```diff
@@ -0,0 +1,5 @@
+function isRequired(value: string) {
```
Comment (roseaugusto) [inline resources/js/utils/validation/isRequired.ts]: put in helper file

File: resources/js/utils/validation/isValidEmail.ts [hunk 16]
```diff
@@ -0,0 +1,7 @@
+function isValidEmail(email: string) {
```
Comment (roseaugusto) [inline resources/js/utils/validation/isValidEmail.ts]: put in helper file

File: resources/js/utils/validation/isValidName.ts [hunk 17]
```diff
@@ -0,0 +1,5 @@
+function isValidName(name: string) {
```
Comment (roseaugusto) [inline resources/js/utils/validation/isValidName.ts]: put in helper file

**Comment by roseaugusto** [comment]
@brixtuyor please add a validation in login and sign up pages. On submit, it won't validate.

https://user-images.githubusercontent.com/70557829/173290278-729d605f-de4e-4247-b285-b35db2508510.mov

File: resources/js/components/Login.tsx (hunk 10 above)
Comment (brixtuyor) [reply resources/js/components/Login.tsx]: Done

File: resources/js/components/Signup.tsx (hunk 11 above)
Comment (brixtuyor) [reply resources/js/components/Signup.tsx]: Done

File: resources/js/utils/validation/isAlphaNumeric.ts (hunk 13 above)
Comment (brixtuyor) [reply resources/js/utils/validation/isAlphaNumeric.ts]: Done

File: resources/js/utils/validation/isMinimum.ts (hunk 14 above)
Comment (brixtuyor) [reply resources/js/utils/validation/isMinimum.ts]: Done

File: resources/js/utils/validation/isRequired.ts (hunk 15 above)
Comment (brixtuyor) [reply resources/js/utils/validation/isRequired.ts]: Done

File: resources/js/utils/validation/isValidEmail.ts (hunk 16 above)
Comment (brixtuyor) [reply resources/js/utils/validation/isValidEmail.ts]: Done

File: resources/js/utils/validation/isValidName.ts (hunk 17 above)
Comment (brixtuyor) [reply resources/js/utils/validation/isValidName.ts]: Done

**Comment by brixtuyor** [comment]
> @brixtuyor please add a validation in login and sign up pages. On submit, it won't validate.
> 
>  Screen.Recording.2022-06-13.at.2.06.50.PM.mov

Fixed

**Review by roseaugusto — APPROVED** [review]
Approved!

**Pull request merged**